## <span style="color:red">Homework 2 </span>

**I- Reading Assignment**

[James, G., Witten, D., Hastie, T., Tibshirani, R., & Taylor, J. (2023). An Introduction to Statistical Learning : With Applications in Python (1st ed. 2023 edition). Springer.](https://www.statlearning.com/)

- Read Chapter 8, Section 1 (8.1: pages 331-342)


In [46]:
import os
import pandas as pd

import jupyter_black

jupyter_black.load()

# ignore warning
import warnings

warnings.filterwarnings("ignore")

**II- Binary Classification**

1- Select your binary outcome

Amongst your categorical variables, select a binary variable to be used as an outcome.
It you have no binary variable, then create a binary variable from one of your categorical variables.

2- Divide your dataset into a test sample and a training sample with a ratio of 20% vs 80% respectively. <br>
Be careful to do so that your outcomes remain representative of the initial data set and well shuffled.

3- Select one classifier and fit it on the training set then estimate its score both on the training set and the test set. <br>Use the accuracy, the ROC AUC, the F1 score and the Kappa to estimate the performance of your classifier.

4- Repeat the same code on two other different classifiers: 

- Recommendation: create a function that does all the job and run it using three different classifiers
. 
- Which is the best classifier ?

In [47]:
os.getcwd()

'/Users/fulin/Documents/GitHub/EMLyon_DSAIS/ML/homework01/data'

In [48]:
PATH = "/Users/fulin/Documents/GitHub/EMLyon_DSAIS/ML/homework01/data"
os.chdir(PATH)

In [72]:
df = pd.read_csv("customer_data_cbn.csv")

In [73]:
cbn_df = df.copy()

In [89]:
cbn_df.head(5)

account_length  number_vmail_messages  total_day_minutes  total_day_calls  \
0           101.0                    0.0               70.9            123.0   
1           137.0                    0.0              223.6             86.0   
2           103.0                   29.0              294.7             95.0   
3            99.0                   33.2              216.8            123.0   
4           108.0                    0.0              197.4             78.0   

   total_day_charge  total_eve_minutes  total_eve_calls  total_eve_charge  \
0            12.050             270.26             73.0            22.976   
1            37.912             244.80            139.0            20.810   
2            40.878             237.30            105.0            20.170   
3            36.860             126.40             88.0            10.740   
4            33.560             124.00            101.0            10.540   

   total_night_minutes  total_night_calls  total_night_charge  \
0                236.0               73.0               10.62   
1                 94.2               81.0                4.24   
2                300.3              127.0               13.51   
3                220.6               82.0                9.93   
4                204.5              107.0                9.20   

   total_intl_minutes  total_intl_calls  total_intl_charge  \
0                10.6               3.0              2.860   
1                 9.5               7.0              2.570   
2                13.7               4.8              3.700   
3                 9.0               2.0              2.432   
4                 7.7               4.0              2.080   

   number_customer_service_calls state  area_code international_plan  \
0                            3.0    HI      510.0                 no   
1                            0.0    MT      510.0                 no   
2                            1.0    OH      408.0                 no   
3                            1.0    NM      415.0                 no   
4                            2.0    SC      415.0                 no   

  voice_mail_plan churn  
0              no    no  
1              no    no  
2             yes    no  
3              no    no  
4              no    no

In [74]:
cbn_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12873 entries, 0 to 12872
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   account_length                 12873 non-null  float64
 1   number_vmail_messages          12873 non-null  float64
 2   total_day_minutes              12873 non-null  float64
 3   total_day_calls                12873 non-null  float64
 4   total_day_charge               12873 non-null  float64
 5   total_eve_minutes              12873 non-null  float64
 6   total_eve_calls                12873 non-null  float64
 7   total_eve_charge               12873 non-null  float64
 8   total_night_minutes            12873 non-null  float64
 9   total_night_calls              12873 non-null  float64
 10  total_night_charge             12873 non-null  float64
 11  total_intl_minutes             12873 non-null  float64
 12  total_intl_calls               12873 non-null 

In [75]:
predictors = cbn_df.drop(columns=["churn"], axis=1)
target_ready = cbn_df["churn"].map({"yes": 1, "no": 0})

In [76]:
predictors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12873 entries, 0 to 12872
Data columns (total 19 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   account_length                 12873 non-null  float64
 1   number_vmail_messages          12873 non-null  float64
 2   total_day_minutes              12873 non-null  float64
 3   total_day_calls                12873 non-null  float64
 4   total_day_charge               12873 non-null  float64
 5   total_eve_minutes              12873 non-null  float64
 6   total_eve_calls                12873 non-null  float64
 7   total_eve_charge               12873 non-null  float64
 8   total_night_minutes            12873 non-null  float64
 9   total_night_calls              12873 non-null  float64
 10  total_night_charge             12873 non-null  float64
 11  total_intl_minutes             12873 non-null  float64
 12  total_intl_calls               12873 non-null 

In [77]:
quanti_features = predictors.iloc[:, :14].columns.tolist()
cat_features = predictors.iloc[:, 15:].columns.tolist()
cat_ohe_features = predictors.iloc[:, 16:].columns.tolist()
cat_lbe_features = ["state"]

This is a place where I carry on struggling, the column state contains the abbreviation of the state, in string format. But before I can train the model I need to do a numerical operation on him. Since he has more than 50 different categories in it, it will cause a serious dimension explosion if I use solo hot coding. So I was thinking of using pca or dimensionality reduction, but this would result in a less interpretable model and no guarantee of having the same representation of the new data in the test set and the training set.
So I am again considering me supervised learning of the state columns, for example performing k-means clustering, setting the number of clusters, and then coding using the classification obtained from the unsupervised learning model as the basis for the classification. This scenario I think would result in a very serious data leakage.
So I am thinking again that I can only use labeled coding, but I am worried about linearity issues with non-tree models. But I had no other choice, so I added the label encoder to the pipeline of my column converter. But a new problem arose, in order to avoid data leakage, I did the column converter after dividing the training and test sets, so I couldn't guarantee that the same states were encoded consistently in my training and test sets, and there was also the possibility that some states existed in only one dataset. So in the end, I chose to manually encode the labels before dividing the training and test sets, with the help of the pandas library.

In [78]:
predictors["state"] = predictors["state"].astype("category").cat.codes

In [79]:
predictors["state"]

0        11
1        26
2        35
3        32
4        40
         ..
12868    26
12869    26
12870    13
12871     0
12872    44
Name: state, Length: 12873, dtype: int8

In [80]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    predictors,
    target_ready,
    test_size=0.2,
    shuffle=True,
    random_state=42,
    stratify=target_ready,
)

In [81]:
X_train_state = X_train["state"]
X_test_state = X_test["state"]

In [82]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler, OneHotEncoder, LabelEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), quanti_features),
        ("cat_ohe", OneHotEncoder(), cat_ohe_features),
    ],
    remainder="drop",
)

pipeline = Pipeline([("preprocessor", preprocessor)])

In [83]:
import numpy as np

X_train = np.column_stack([pipeline.fit_transform(X_train), X_train_state])
display(X_train.shape)
X_test = np.column_stack([pipeline.fit_transform(X_test), X_test_state])
display(X_test.shape)

(10298, 22)

(2575, 22)

In [84]:
from sklearn.metrics import accuracy_score, roc_auc_score, f1_score, cohen_kappa_score
import time


def evaluate_classifier(clf, X_train, y_train, X_test, y_test):
    start_time = time.time()

    clf.fit(X_train, y_train)

    end_time = time.time()
    training_time = end_time - start_time

    y_train_pred = clf.predict(X_train)
    y_test_pred = clf.predict(X_test)

    metrics = {
        "Accuracy (train)": accuracy_score(y_train, y_train_pred),
        "Accuracy (test)": accuracy_score(y_test, y_test_pred),
        "ROC AUC (train)": roc_auc_score(y_train, y_train_pred),
        "ROC AUC (test)": roc_auc_score(y_test, y_test_pred),
        "F1 Score (train)": f1_score(y_train, y_train_pred),
        "F1 Score (test)": f1_score(y_test, y_test_pred),
        "Kappa (train)": cohen_kappa_score(y_train, y_train_pred),
        "Kappa (test)": cohen_kappa_score(y_test, y_test_pred),
        "Training Time (s)": training_time,
    }
    return metrics

In [85]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier

classifiers = [
    ("Logistic Regression", LogisticRegression(max_iter=100)),
    ("KNN", KNeighborsClassifier()),
    ("Decision Trees", DecisionTreeClassifier()),
    ("Random Forest", RandomForestClassifier()),
    ("SVM", SVC(probability=True)),
    ("XGBooster", XGBClassifier(use_label_encoder=False, eval_metric="logloss")),
]

In [86]:
results = {}
for name, clf in classifiers:
    results[name] = evaluate_classifier(clf, X_train, y_train, X_test, y_test)

In [87]:
display(results)

{'Logistic Regression': {'Accuracy (train)': 0.8651194406680909,
  'Accuracy (test)': 0.8710679611650486,
  'ROC AUC (train)': 0.5626020420085276,
  'ROC AUC (test)': 0.5760856059363522,
  'F1 Score (train)': 0.22790439132851587,
  'F1 Score (test)': 0.2654867256637168,
  'Kappa (train)': 0.18400920869797077,
  'Kappa (test)': 0.22270229379652962,
  'Training Time (s)': 0.15522098541259766},
 'KNN': {'Accuracy (train)': 0.9584385317537386,
  'Accuracy (test)': 0.9013592233009708,
  'ROC AUC (train)': 0.8757655358290836,
  'ROC AUC (test)': 0.7096206032773197,
  'F1 Score (train)': 0.8381240544629349,
  'F1 Score (test)': 0.5590277777777778,
  'Kappa (train)': 0.8145674991545708,
  'Kappa (test)': 0.5078126881306138,
  'Training Time (s)': 0.00038886070251464844},
 'Decision Trees': {'Accuracy (train)': 1.0,
  'Accuracy (test)': 0.9398058252427185,
  'ROC AUC (train)': 1.0,
  'ROC AUC (test)': 0.8754442075337597,
  'F1 Score (train)': 1.0,
  'F1 Score (test)': 0.7867950481430535,
  'Kap

In [88]:
from prettytable import PrettyTable

train_table = PrettyTable()
train_table.title = "Training Set Performance"
train_table.field_names = [
    "Classifier",
    "Accuracy",
    "ROC AUC",
    "F1 Score",
    "Kappa",
    "Training Time (s)",
]

for name, metrics in results.items():
    train_table.add_row(
        [
            name,
            f"{metrics['Accuracy (train)']:.4f}",
            f"{metrics['ROC AUC (train)']:.4f}",
            f"{metrics['F1 Score (train)']:.4f}",
            f"{metrics['Kappa (train)']:.4f}",
            f"{metrics['Training Time (s)']:.4f}",
        ]
    )

test_table = PrettyTable()
test_table.title = "Test Set Performance"
test_table.field_names = [
    "Classifier",
    "Accuracy",
    "ROC AUC",
    "F1 Score",
    "Kappa",
]

for name, metrics in results.items():
    test_table.add_row(
        [
            name,
            f"{metrics['Accuracy (test)']:.4f}",
            f"{metrics['ROC AUC (test)']:.4f}",
            f"{metrics['F1 Score (test)']:.4f}",
            f"{metrics['Kappa (test)']:.4f}",
        ]
    )
print(train_table)
print(test_table)

+----------------------------------------------------------------------------------+
|                             Training Set Performance                             |
+---------------------+----------+---------+----------+--------+-------------------+
|      Classifier     | Accuracy | ROC AUC | F1 Score | Kappa  | Training Time (s) |
+---------------------+----------+---------+----------+--------+-------------------+
| Logistic Regression |  0.8651  |  0.5626 |  0.2279  | 0.1840 |       0.1552      |
|         KNN         |  0.9584  |  0.8758 |  0.8381  | 0.8146 |       0.0004      |
|    Decision Trees   |  1.0000  |  1.0000 |  1.0000  | 1.0000 |       0.1096      |
|    Random Forest    |  1.0000  |  1.0000 |  1.0000  | 1.0000 |       1.3955      |
|         SVM         |  0.8585  |  0.5000 |  0.0000  | 0.0000 |       5.4646      |
|      XGBooster      |  0.9989  |  0.9962 |  0.9962  | 0.9956 |       0.1130      |
+---------------------+----------+---------+----------+--------+-

### Model Analysis based on Performance and Computational Cost

#### 1. Computational Cost vs. Performance:
- The **SVM** has a considerably longer training time compared to the other models. Despite this high computational cost, its performance on the test set isn't top-tier. This highlights the importance of balancing model performance with computational expenses.
  
- On the other hand, the **Random Forest** exhibits exceptional performance on the test set but also comes with a lengthier training time. Given its superior performance, this computational cost might be deemed acceptable.

#### 2. Overfitting:
- The **Decision Trees** model showcases perfect performance on the training set. However, a drop in performance on the test set signals a classic case of overfitting.
  
- Both **Random Forest** and **XGBooster** shine on the training set, but they manifest a smaller performance drop on the test set. This suggests that they might be generalizing better than some of the other models.

#### 3. Model Selection:
- When factoring in both performance and computational cost, the **Random Forest** seems like an optimal choice. It stands out in terms of test set performance, although it requires more training time compared to some models.
  
- The near-zero training time for **KNN** is noteworthy. This is due to the nature of KNN being an instance-based learning method, with its primary computational cost being during the prediction phase rather than the training phase.

In conclusion, selecting the ideal model requires a holistic view that encompasses multiple factors, including model performance, computational expenses, business requirements, and other specific contextual considerations.

**III- Multiclass Classification** (Optional with good bonus point)

This time, select a categorical variable having more than 2 categories. I you have none, then categorize one of your quantitative data (e.g in using `pd.cut()` and the corresponding quantiles) into at least three classes.

Copy the codes of PART I and adjust to the new outcome. Use only Accuracy and F1 Score for performance (or sklearn `classification_report`)

In [90]:
df_multi = df.copy()

In [91]:
df_multi["state"].unique()

array(['HI', 'MT', 'OH', 'NM', 'SC', 'IA', 'ND', 'LA', 'MO', 'TX', 'AR',
       'ME', 'DE', 'MN', 'KS', 'NC', 'NV', 'CO', 'TN', 'OR', 'NE', 'ID',
       'WY', 'MD', 'AZ', 'VT', 'MS', 'CA', 'WI', 'SD', 'WV', 'MI', 'UT',
       'NJ', 'MA', 'AL', 'FL', 'VA', 'IL', 'PA', 'RI', 'DC', 'IN', 'CT',
       'GA', 'NY', 'NH', 'KY', 'WA', 'AK', 'OK'], dtype=object)

In [92]:
# Define the states for each region
west_states = ["WA", "OR", "CA", "NV", "ID", "UT", "AZ", "MT", "WY", "CO", "NM"]
mid_states = ["ND", "SD", "NE", "KS", "MO", "IA", "MN", "WI", "IL", "IN", "OH", "MI"]


# Create a new column 'region' based on the 'state' column
def state_to_region(state):
    if state in west_states:
        return "West"
    elif state in mid_states:
        return "Mid"
    else:
        return "East"


df_multi["region"] = df_multi["state"].apply(state_to_region)

# Display the distribution of the new 'region' column
df_multi["region"].value_counts()

East    7019
Mid     3058
West    2796
Name: region, dtype: int64

In [93]:
from sklearn.metrics import classification_report
from sklearn.preprocessing import LabelEncoder

# Prepare the predictors and target
predictors_new = df_multi.drop(columns=["region", "state", "churn"])
target_new = df_multi["region"]

# Encode the target variable since many algorithms require numerical input
label_encoder = LabelEncoder()
target_encoded = label_encoder.fit_transform(target_new)

# Split the data into training and testing sets
X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(
    predictors_new,
    target_encoded,
    test_size=0.2,
    shuffle=True,
    random_state=42,
    stratify=target_encoded,
)

# Update the preprocessor to exclude 'state'
quanti_features_new = predictors_new.iloc[:, :13].columns.tolist()
cat_ohe_features_new = predictors_new.iloc[:, 15:].columns.tolist()

preprocessor_new = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), quanti_features_new),
        ("cat_ohe", OneHotEncoder(), cat_ohe_features_new),
    ],
    remainder="drop",
)

pipeline_new = Pipeline([("preprocessor", preprocessor_new)])

# Preprocess the data
X_train_processed = pipeline_new.fit_transform(X_train_new)
X_test_processed = pipeline_new.transform(X_test_new)

# Evaluate the classifiers
results_new = {}
for name, clf in classifiers:
    clf.fit(X_train_processed, y_train_new)
    y_train_pred_new = clf.predict(X_train_processed)
    y_test_pred_new = clf.predict(X_test_processed)

    print(f"Classifier: {name}")
    print("Training Set Performance:")
    print(
        classification_report(
            y_train_new, y_train_pred_new, target_names=label_encoder.classes_
        )
    )
    print("Test Set Performance:")
    print(
        classification_report(
            y_test_new, y_test_pred_new, target_names=label_encoder.classes_
        )
    )
    print("-" * 80)

Classifier: Logistic Regression
Training Set Performance:
              precision    recall  f1-score   support

        East       0.55      1.00      0.71      5615
         Mid       0.00      0.00      0.00      2446
        West       0.00      0.00      0.00      2237

    accuracy                           0.55     10298
   macro avg       0.18      0.33      0.24     10298
weighted avg       0.30      0.55      0.38     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.55      1.00      0.71      1404
         Mid       0.00      0.00      0.00       612
        West       0.00      0.00      0.00       559

    accuracy                           0.55      2575
   macro avg       0.18      0.33      0.24      2575
weighted avg       0.30      0.55      0.38      2575

--------------------------------------------------------------------------------
Classifier: KNN
Training Set Performance:
              precision    recall  

Classifier: Logistic Regression
Training Set Performance:
              precision    recall  f1-score   support

        East       0.55      1.00      0.71      5615
         Mid       0.00      0.00      0.00      2446
        West       0.00      0.00      0.00      2237

    accuracy                           0.55     10298
   macro avg       0.18      0.33      0.24     10298
weighted avg       0.30      0.55      0.38     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.55      1.00      0.71      1404
         Mid       0.00      0.00      0.00       612
        West       0.00      0.00      0.00       559

    accuracy                           0.55      2575
   macro avg       0.18      0.33      0.24      2575
weighted avg       0.30      0.55      0.38      2575

--------------------------------------------------------------------------------
Classifier: KNN
Training Set Performance:
              precision    recall  f1-score   support

        East       0.86      0.93      0.89      5615
         Mid       0.86      0.80      0.83      2446
        West       0.86      0.76      0.81      2237

    accuracy                           0.86     10298
   macro avg       0.86      0.83      0.84     10298
weighted avg       0.86      0.86      0.86     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.70      0.82      0.75      1404
         Mid       0.57      0.51      0.54       612
        West       0.62      0.42      0.50       559

    accuracy                           0.66      2575
   macro avg       0.63      0.58      0.60      2575
weighted avg       0.65      0.66      0.65      2575

--------------------------------------------------------------------------------
Classifier: Decision Trees
Training Set Performance:
              precision    recall  f1-score   support

        East       1.00      1.00      1.00      5615
         Mid       1.00      1.00      1.00      2446
        West       1.00      1.00      1.00      2237

    accuracy                           1.00     10298
   macro avg       1.00      1.00      1.00     10298
weighted avg       1.00      1.00      1.00     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.85      0.82      0.84      1404
         Mid       0.75      0.76      0.75       612
        West       0.70      0.74      0.72       559

    accuracy                           0.79      2575
   macro avg       0.77      0.77      0.77      2575
weighted avg       0.79      0.79      0.79      2575

--------------------------------------------------------------------------------
Classifier: Random Forest
Training Set Performance:
              precision    recall  f1-score   support

        East       1.00      1.00      1.00      5615
         Mid       1.00      1.00      1.00      2446
        West       1.00      1.00      1.00      2237

    accuracy                           1.00     10298
   macro avg       1.00      1.00      1.00     10298
weighted avg       1.00      1.00      1.00     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.85      0.99      0.92      1404
         Mid       0.99      0.79      0.88       612
        West       0.99      0.79      0.88       559

    accuracy                           0.90      2575
   macro avg       0.94      0.86      0.89      2575
weighted avg       0.91      0.90      0.90      2575

--------------------------------------------------------------------------------
Classifier: SVM
Training Set Performance:
              precision    recall  f1-score   support

        East       0.57      1.00      0.72      5615
         Mid       0.96      0.10      0.18      2446
        West       0.95      0.07      0.13      2237

    accuracy                           0.58     10298
   macro avg       0.83      0.39      0.34     10298
weighted avg       0.75      0.58      0.47     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.56      1.00      0.72      1404
         Mid       0.89      0.05      0.10       612
        West       0.86      0.04      0.09       559

    accuracy                           0.57      2575
   macro avg       0.77      0.36      0.30      2575
weighted avg       0.70      0.57      0.43      2575

--------------------------------------------------------------------------------
Classifier: XGBooster
Training Set Performance:
              precision    recall  f1-score   support

        East       0.95      1.00      0.97      5615
         Mid       0.99      0.93      0.96      2446
        West       0.99      0.92      0.96      2237

    accuracy                           0.97     10298
   macro avg       0.98      0.95      0.96     10298
weighted avg       0.97      0.97      0.97     10298

Test Set Performance:
              precision    recall  f1-score   support

        East       0.78      0.97      0.87      1404
         Mid       0.92      0.67      0.78       612
        West       0.90      0.61      0.73       559

    accuracy                           0.82      2575
   macro avg       0.87      0.75      0.79      2575
weighted avg       0.84      0.82      0.81      2575

--------------------------------------------------------------------------------


Logistic regression and SVM perform relatively poorly on this multi-class classification problem.
Decision trees and random forests both perform 100% on the training set, but the performance on the test set drops slightly, which could be a sign of overfitting.
XGBooster performs well on both the training and test sets and seems to be a good choice for this multi-class classification problem.